File: cpp/common/TreeNode.cc
#include <iostream>
#include <queue>
#include <string>
#include "TreeNode.h"

TreeNode::TreeNode() : val(0), left(nullptr), right(nullptr) {}
TreeNode::TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
TreeNode::TreeNode(int x, TreeNode *left, TreeNode *right) :
	val(x), left(left), right(right) {}

namespace util {
	// LeetCode Tree Visualizer에 맞추어서 구현함. 디버깅용.
	TreeNode *makeTree(vector<int> nums) {
		if (nums.empty()) return nullptr;

		queue<TreeNode *> queue;
		int index = 1;
		TreeNode *root = new TreeNode(nums[0]);
		queue.push(root);

		while (index < nums.size()) {
			TreeNode *cur = queue.front();
			queue.pop();

			cur->left = nums[index] == NONE ? nullptr : new TreeNode(nums[index]);
			index++;

			if (index < nums.size()) {
				cur->right = nums[index] == NONE ? nullptr : new TreeNode(nums[index]);
				index++;
			}

			if (cur->left != nullptr) queue.push(cur->left);
			if (cur->right != nullptr) queue.push(cur->right);
		}

		return root;
	}

	void printTree(TreeNode *root) {
		if (root == nullptr) {
			cout << "no element" << endl;
			return;
		}
		
		queue<TreeNode *> queue;
		string str = "";
		queue.push(root);

		str += "[";
		if (root->left == nullptr && root->right == nullptr) {
			str += to_string(root->val) + "]";
		} else {
			str += to_string(root->val) + ", ";
		}

		while (!queue.empty()) {
			TreeNode *cur = queue.front();
			TreeNode *left = cur->left;
			TreeNode *right = cur->right;
			string value;
			queue.pop();

			if (left == nullptr && right == nullptr) continue;

			value = left == nullptr ? "NONE" : to_string(left->val);
			str += value + ", ";

			value = right == nullptr ? "NONE" : to_string(right->val);
			str += value + ", ";

			if (left != nullptr) queue.push(left);
			if (right != nullptr) queue.push(right);
		}

		str.pop_back();
		str.pop_back();
		str += "]";
		cout << str << endl;
	}
}

File: python/src/common/tree_node.py
from typing import List, Optional
from queue import Queue


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    @staticmethod
    def make(nums: List[Optional[int]]):
        if len(nums) == 0 or nums[0] is None:
            return None

        root = TreeNode(nums[0])
        queue = Queue()
        queue.put(root)

        for index in range(1, len(nums), 2):
            node = queue.get()

            if nums[index] is not None:
                node.left = TreeNode(nums[index])
            if index + 1 < len(nums) and nums[index + 1] is not None:
                node.right = TreeNode(nums[index + 1])

            if node.left is not None:
                queue.put(node.left)
            if node.right is not None:
                queue.put(node.right)

        return root
